py/desispec/validredshifts.py

```diff
@@ -0,0 +1,190 @@
+            if name in ['BGS_FAINT', 'BGS_BRIGHT']:
+                cat[name] = cat['BGS_TARGET'] & bgs_mask[name] > 0
+            else:
+                cat[name] = cat['DESI_TARGET'] & desi_mask[name] > 0
+        # # Bitmask definitions: https://github.com/desihub/desitarget/blob/master/py/desitarget/data/targetmask.yaml
+        # cat['LRG'] = cat['DESI_TARGET'] & 2**0 > 0
+        # cat['ELG'] = cat['DESI_TARGET'] & 2**1 > 0
+        # cat['QSO'] = cat['DESI_TARGET'] & 2**2 > 0
+        # cat['ELG_LOP'] = cat['DESI_TARGET'] & 2**5 > 0
+        # cat['ELG_HIP'] = cat['DESI_TARGET'] & 2**6 > 0
+        # cat['ELG_VLO'] = cat['DESI_TARGET'] & 2**7 > 0
+        # cat['BGS_ANY'] = cat['DESI_TARGET'] & 2**60 > 0
+        # cat['BGS_FAINT'] = cat['BGS_TARGET'] & 2**0 > 0
+        # cat['BGS_BRIGHT'] = cat['BGS_TARGET'] & 2**1 > 0
+
+    cat = actually_validate(cat, fiberstatus_cut=fiberstatus_cut, ignore_emline=ignore_emline)
+    cat = cat[output_columns]
+
+    return cat
+
+
+def actually_validate(cat, fiberstatus_cut=True, ignore_emline=False):
+    '''
+    Apply redshift quality criteria
+
+    Args:
+        cat: astropy table with the necessary columns for redshift quality determination
+
+    Options:
+        fiberstatus_cut: bool (default True), if True, impose requirements on COADD_FIBERSTATUS and ZWARN
+        return_target_columns: bool (default False), if True, include columns that indicate if the object belongs to each class of DESI targets
+        extra_columns: list of str (default None), additional columns to include in the output
+        emline_path: str (default None), specify the location of the emline file; by default the emline file is in the same directory as the redrock file
+        ignore_emline: bool (default False), if True, ignore the emline file and do not validate the ELG redshift
+
+    Returns:
+        cat: astropy table with boolean columns (e.g., GOOD_BGS) added
+    '''
+
+    # BGS
+    cat['GOOD_BGS'] = cat['ZWARN']==0
+    cat['GOOD_BGS'] &= cat['DELTACHI2']>40
+    if fiberstatus_cut:
+        cat['GOOD_BGS'] &= get_good_fiberstatus(cat)
+
+    # LRG
+    cat['GOOD_LRG'] = cat['ZWARN']==0
+    cat['GOOD_LRG'] &= cat['Z']<1.5
+    cat['GOOD_LRG'] &= cat['DELTACHI2']>15
+    if fiberstatus_cut:
+        cat['GOOD_LRG'] &= get_good_fiberstatus(cat)
+
+    # ELG
+    if not ignore_emline:
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore')
+            cat['GOOD_ELG'] = (cat['OII_FLUX']>0) & (cat['OII_FLUX_IVAR']>0)
+            cat['GOOD_ELG'] &= np.log10(cat['OII_FLUX'] * np.sqrt(cat['OII_FLUX_IVAR'])) > 0.9 - 0.2 * np.log10(cat['DELTACHI2'])
+        if fiberstatus_cut:
+            cat['GOOD_ELG'] &= get_good_fiberstatus(cat)
+
+    # QSO - adopted from the code from Edmond
+    # https://github.com/echaussidon/LSS/blob/8ca53f4c38cfa29722ee6958687e188cc894ed2b/py/LSS/qso_cat_utils.py#L282
```

Comment:
Thanks for linking to this LSS qso_cat_utils code with the same bits, but are these defined anywhere as a bitmask object, yaml/json config file, or something that maps name <-> bitnumber ?  Is there a repo that "owns" the concept of QSO_MASKBITS?  Having bit definitions buried in code is problematic for long term maintenance.

Reply:
@echaussidon provided the script and may be able to answer this. My thinking is that QSO_MASKBITS is only an intermediate value and not part of the output, and its definition is quite simple (it's only those few lines), so perhaps it's not so critical to have it as an object or yaml/json file.

Reply:
@rongpu, @sbailey: QSO_MASKBITS is to track from which criteria the QSO is selected. There is no yaml file for the moment. I can create one (like the desitarget maskbit) and add it in the LSS repo.

Reply:
@sbailey could you advise what should be done here? E.g., I could simply remove QSO_MASKBITS and rewrite it as a series of AND statements.